pkg/controller/logforwarding/logforwarding_controller.go

```diff
@@ -0,0 +1,134 @@
+	}
+
+	// Watch for changes to primary resource LogForwarding
+	err = c.Watch(&source.Kind{Type: &loggingv1alpha1.LogForwarding{}}, &handler.EnqueueRequestForObject{}, pred)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileLogForwarding implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileLogForwarding{}
+
+// ReconcileLogForwarding reconciles a LogForwarding object
+type ReconcileLogForwarding struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a LogForwarding object and makes changes based on the state read
+// and what is in the LogForwarding.Spec
+//
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileLogForwarding) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling LogForwarding")
+
+	// Fetch the LogForwarding instance
+	instance := &loggingv1alpha1.LogForwarding{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Mark the old LogForwarding CR as obsolete before converting to a ClusterLogForwarder CR
+	instance.Status = &loggingv1alpha1.ForwardingStatus{
+		State:  loggingv1alpha1.LogForwardingStateObsolete,
+		Reason: loggingv1alpha1.LogForwardingReasonObsolete,
+		Message: fmt.Sprintf(
+			"LogForwarding resource %q is replaced by ClusterLogForwarder resource `instance`",
+			instance.GetName(),
+		),
+	}
+
+	updateOpts := []client.UpdateOption{}
+	err = r.client.Status().Update(context.TODO(), instance, updateOpts...)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	clf := &loggingv1.ClusterLogForwarder{}
+	if err := instance.ConvertTo(clf); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	createOpts := []client.CreateOption{}
+	err = r.client.Create(context.TODO(), clf, createOpts...)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			// ClusterLogForwarder already exists, don't requeue
+			reqLogger.Info("Skip reconcile: ClusterLogFowarder already exists")
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
```

Comment:
shouldn't we requeue the event to make sure we eventually create the GA LF object?

Reply:
AFAIK, returning an err will always requeue:
```
// The Controller will requeue the Request to be processed again if the returned error is non-nil or
// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
```